### PR TITLE
Adding shunting yard (Reverse_Polish) algorithm under Algorithm/String

### LINKED ISSUE #195: Shunting Yard Algorithm
I was wondering if you wanted to add Shunting-yard algorithm. If yes, I can do it.

### PR DESCRIPTION
**Pull request by DevRin**
Shunting yard algorithm has been added. Three files are uploaded under:
1- reverse_polish.cpp under algorithm/string
2- reverse_polish.hpp under include/algorithm/string
3- reverse_polish.test.cpp under test/algorithm/string

- [x] Read the [contribution guidelines][contrib-guidelines]
- [x] Added [unit tests][unit-tests] (or unit tests have already been added)
- [ ] Updated the [Contents][contents]

closes #195

[contrib-guidelines]: https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md
[unit-tests]: https://github.com/faheel/Algos/blob/master/C%2B%2B/UNIT_TESTS.md
[contents]: https://github.com/faheel/Algos/tree/master/C%2B%2B#contents

### CODE REVIEW

File: C++/test/algorithm/string/reverse_polish.test.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+	result_vector = test_operations.to_reverse_polish(input_vector);
+
+	EXPECT_EQ(result_vector.at(0), "hello");
+	EXPECT_EQ(result_vector.at(1), "he2o");
+	EXPECT_EQ(result_vector.at(2), "hell0");
+
+}
+
+int main(int argc, char **argv) {
```
Comment (alxmjo) [inline C++/test/algorithm/string/reverse_polish.test.cpp:61]: No main() necessary.

File: C++/test/algorithm/string/reverse_polish.test.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+TEST_F(OperatorOperationsTest,emptyCase){
+
+	input_vector.push_back(" ");
+	input_vector.push_back("");
+
+	result_vector = test_operations.to_reverse_polish(input_vector);
+
+	EXPECT_EQ(result_vector.at(0), "");
+	EXPECT_EQ(result_vector.at(1), "");
+
```
Comment (alxmjo) [inline C++/test/algorithm/string/reverse_polish.test.cpp:43]: Remove extra newlines.

File: C++/test/algorithm/string/reverse_polish.test.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	result_vector = test_operations.to_reverse_polish(input_vector);
+
+	EXPECT_EQ(result_vector.at(0), "342*15-23^^/+");
+	EXPECT_EQ(result_vector.at(1), "342*15-23^^/+");
+	EXPECT_EQ(result_vector.at(2), "31^56*/10+");
+
+
+}
+TEST_F(OperatorOperationsTest,emptyCase){
```
Comment (alxmjo) [inline C++/test/algorithm/string/reverse_polish.test.cpp:34]: Space between arguments.

File: C++/include/algorithm/string/reverse_polish.hpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#include <stack>
+#include <unordered_map>
+#include <algorithm>
+
+
+class OperatorOperations {
+ public:
+	std::vector <std::string> to_reverse_polish (const std::vector <std::string> &);
+	static std::unordered_map<char, int> operator_precedence;
+    static std::unordered_map<char, std::string> operator_association;
```
Comment (alxmjo) [inline C++/include/algorithm/string/reverse_polish.hpp:16]: Fix justification.

Comment (alxmjo) [inline C++/include/algorithm/string/reverse_polish.hpp:16]: If you add `using std::string` at the top then you won't have to preface all of your strings with std::.

File: C++/source/string/reverse_polish.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
+/*
+ <Shunting Yard Algorithm>
```
Comment (alxmjo) [inline C++/source/string/reverse_polish.cpp:2]: No angle brackets necessary around names and comments.

File: C++/source/string/reverse_polish.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+-------------------
```
Comment (alxmjo) [inline C++/source/string/reverse_polish.cpp:3]: Make sure hyphens underneath titles are same length as titles.

File: C++/source/string/reverse_polish.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+---------------
+<O(N*M*P) where N is the size of input vector, i.e. equations, M is the size of the greatest string in the vector, and P is maximum number of operators in the string>
+
+Space complexity
+----------------
+<O(N+P) = O(N) : where N is the size of input vector, i.e. equations, and  P is maximum number of operators in the string, i.e. stack operations. >
+*/
+
+#include "reverse_polish.hpp"
+using namespace std;
```
Comment (alxmjo) [inline C++/source/string/reverse_polish.cpp:17]: Don't use `using namespace std`. Instead, reference the individual items, as in `using std::string`.

File: C++/source/string/reverse_polish.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+            else {
+                if (equation[j] == '(') {
+                    operators.push(equation[j]);
+                }
+                else {
+                    if (equation[j] == ')') {
+
+                        while (operators.top() != '(') {
+
```
Comment (alxmjo) [inline C++/source/string/reverse_polish.cpp:53]: Fix unnecessary white space.

File: C++/source/string/reverse_polish.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                            operators.pop();
+                        }
+                        operators.pop();
+                    }
+                    else {
+                        if (operators.empty()) {
+                            operators.push(equation[j]);
+                        }
+                        else {
+                            while (!operators.empty() && (operators.top() != '(') && (operations.operator_precedence[operators.top()] > operations.operator_precedence[equation[j]] || (operations.operator_precedence[operators.top()] == operations.operator_precedence[equation[j]] && operations.operator_association[operators.top()] == "left"))) {
```
Comment (alxmjo) [inline C++/source/string/reverse_polish.cpp:64]: This is a lot for one line. Can you break it up nicely?

File: C++/test/algorithm/string/reverse_polish.test.cpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
+#include "reverse_polish.cpp"
+#include "gtest/gtest.h"
```
Comment (alxmjo) [inline C++/test/algorithm/string/reverse_polish.test.cpp:2]: Please refactor this to use the Catch.hpp testing framework. Look at any other unit test for an example.

File: C++/test/algorithm/string/reverse_polish.test.cpp [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class OperatorOperationsTest : public :: testing::Test{
+
+protected:
+	virtual void SetUp() override {
```
Comment (alxmjo) [inline C++/test/algorithm/string/reverse_polish.test.cpp:8]: Are these necessary?

File: C++/test/algorithm/string/reverse_polish.test.cpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+		//
+	}
+	virtual void TearDown() {
```
Comment (alxmjo) [inline C++/test/algorithm/string/reverse_polish.test.cpp:11]: Is this necessary?

**Review by alxmjo — CHANGES_REQUESTED** [review]
Thanks for your submission, Nasrin. Please have a look at my notes and let me know if you have any questions.

**Comment by DevRin** [comment]
Thanks for you thorough review. yes, I talked with Faeel and I should change the testing framework after my exams to unify. I'll close this request and open a new one after I've done the changes.

**Comment by alxmjo** [comment]
Sounds good. Thanks Nasrin.

**Pull request closed without merging**